src/FlightController.cpp

```diff
@@ -4,106 +4,293 @@
+        msp::msg::RxMap rx_map(fw_variant_);
+        client_.sendMessage(rx_map);
+        std::cout << rx_map;
+        channel_map_ = rx_map.map;
+    }
+    
+    return true;
+}
+
+bool FlightController::disconnect(bool blocking, double timeout)
+{
+    stopMspControl();
+    return client_.disconnect();
+
+}    
+    
+void FlightController::arm(bool blocking) {
+    armed_ = true;
+}
+
+void FlightController::disarm(bool blocking) {
+    armed_ = false;
+}
+
+bool FlightController::armSet() {
+    return armed_;
+}
+
+
+void FlightController::printDebug(bool on)
+{
+    client_.printDebug(on);
+}
+
+
+void FlightController::setFlightMode(FlightMode mode)
+{
+    {
+        std::lock_guard<std::mutex> lock(msp_updates_mutex);
+        flight_mode_ = mode;
+    }
+    generateMSP();
+}
+
+FlightMode FlightController::getFlightMode()
+{
+    return flight_mode_;
+}
+
+void FlightController::setControlSource(ControlSource source)
+{
+    if (source == control_source_) return;
+    //client_.printDebug(true);
+    
+    msp::msg::RxConfig rxConfig(fw_variant_);
+    if (!client_.sendMessage(rxConfig)) std::cout << "client_.sendMessage(rxConfig) failed" << std::endl;
+    std::cout << rxConfig;
+    
+    msp::msg::SetRxConfig setRxConfig(fw_variant_);
+    static_cast<msp::msg::RxConfigSettings&>(setRxConfig) = static_cast<msp::msg::RxConfigSettings&>(rxConfig);
+    
+    if (source == ControlSource::SBUS) {
+        setRxConfig.receiverType = 3;
+        setRxConfig.serialrx_provider = 2;
+    } else if (source == ControlSource::MSP) {
+        setRxConfig.receiverType = 4;
+    }
+    //std::cout << setRxConfig;
+    client_.sendMessage(setRxConfig);
+    
+    if (source == ControlSource::MSP) startMspControl();
+    else stopMspControl();
+    
+    control_source_ = source;
+    //client_.printDebug(false);
+}
+
+ControlSource FlightController::getControlSource()
+{
+    msp::msg::RxConfig rxConfig(fw_variant_);
+    client_.sendMessage(rxConfig);
+    
+    if (rxConfig.receiverType && rxConfig.receiverType() == 4) return ControlSource::MSP;
+    else if (rxConfig.serialrx_provider() == 2) return ControlSource::SBUS;
+    return ControlSource::OTHER;
+    
+}
+
+void FlightController::setRPYT(std::array<double,4>& rpyt)
+{
+    //std::cout << "void FlightController::setRPYT(std::array<double,4>& rpyt)" << std::endl;
+    {
+        std::lock_guard<std::mutex> lock(msp_updates_mutex);
+        rpyt_.swap(rpyt);
     }
+    generateMSP();
+    //std::cout << "finished setRPYT" << std::endl;
 }
 
-bool FlightController::isFirmware(const FirmwareType firmware_type) {
-    return firmware == firmware_type;
+void FlightController::startMspControl()
+{
+    std::cout << "STARTING MSP CONTROL" << std::endl;
+    msp_timer_.start();
+}
+
+void FlightController::stopMspControl()
+{
+    std::cout << "STOPPING MSP CONTROL" << std::endl;
+    msp_timer_.stop();
 }
 
+void FlightController::generateMSP()
+{
+    std::vector<uint16_t> cmds(18,1000);
```

Comment:
Why 18 channels? The maximum amount of channels is 16 with two optional on/off switches. We should only send as many channels as supported by the RC to save some bandwidth.

Reply:
I am still not sure about the 18 channels. I thought there are maximum 16 channels +2 on/off flags. And do we need to send all of them or just those that are actually used?

Reply:
updated to 6.